src/docs/vtable-alternative.md

```diff
@@ -0,0 +1,92 @@
+# Vtable Alternative
+
+In Binding Tools for Swift, we use a simulated vtable in order to delegate implementations of Swift types to C#.
+
+This process works, but it has a number of problems. All of them stem from the same issue: Swift knows nothing about C# and needs to call a function exported from C# that is `[UnmanagedCallersOnly]`. I refer to this type of function as a Receiver. The Receiver then needs to get to the actual C# implementation.
+1. the Swift vtable - a struct of C calling convention function pointers to Receivers
+2. a C# vtable to hold delegates for the Receivers that will be marshalled as unmanaged functions pointers. This definition lives in a separate static class since Receivers can't live inside of a (possibly) generic type
+3. a C# vtable to hold managed delegates to the actual implementation which lives in the bound type.
+
+As you can imagine, it can be painful to keep these types in sync.
+
+C# must call into Swift in order to initialize the vtable. In addition, if the type being bound is generic, then there needs to be a vtable for each unique specialization of the type.
+
+Because the population of the vtable is done at runtime, the calling Swift code is required to test the function pointer for null, even though it should never be null.
+
+Most of these issues can be solved if Swift can call C# exported functions directly. This can be done if the binding library is compiled with NativeAOT, generating a dylib.
```

Comment:
Isn't that only the case if you have `IlcDisableReflection` set?

Reply:
`IlcDisableReflection` is an unsupported experimental option. You do not need to worry about it. It is likely going to be deleted at some point.

However, I do not see what you get from using native AOT. Native AOT compiled dylibs are fully self-contained and each native AOT compiler dylib has its own copy of the runtime, GC, etc. The only way to call code outside of them is through unmanaged interop.

Reply:
Sure - using NativeAOT allows Swift code to call C# code directly instead of C# providing function pointers (ie, a vtable) and because of the way that Swift is, it is a requirement to have null check on each function pointer before calling it and it is a requirement to have an alternative path for the code for the case of the null check failing (which it never will, but there is no way to tell Swift that's the case). Being able to let the linker do the work cuts out the memory overhead of the vtables, which is decidedly small but will grow in generics which require a vtable for each specialization.
Another approach suggested by Rolf today was to write C code that knows about the runtime and calls into it and let Swift link to that.

Reply:
I do not think you want to have a native AOT compiled .dylib in the process in addition the regular .NET runtime. Each native AOT compiled .dylib has its own GC heap, etc. It is a lot of overhead.

If you need to build some glue code, writing it in C sounds like a better option to me.